File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int solve(vector<int>&nums,long long sum) {
        int cnt=1;
        long long sSum=0;
        for(int i=0;i<nums.size();i++) {
            if(sSum+nums[i]>sum) {
                sSum=nums[i];
                cnt++;
            }else {
                sSum+=nums[i];
            }
        }
        return cnt;
    }
    int splitArray(vector<int>& nums, int k) {
        if(k>nums.size()) return -1;
        int l = *max_element(nums.begin(),nums.end());
        long long h = accumulate(nums.begin(),nums.end(),0);
        while(l<=h) {
            int mid = l+(h-l)/2;
            int div = solve(nums,mid);
            if(div>k) l=mid+1;
            else h=mid-1;
        }
        return l;
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int l=0,h=0;
        for(int n: nums) {
            l = Math.max(l,n);
            h+=n;
        }
        int ans=h;
        while(l<=h) {
            int mid = l + (h-l)/2;
            if(canPlace(nums,mid,k)) {
                ans=mid;
                 h=mid-1;
            }else {
               l=mid+1;
            }
        }
        return ans;
    }

    public boolean canPlace(int nums[], int mid, int k) {
        int count = 1;
        int currSum = 0;
        for(int n: nums) {
            if(currSum+n <= mid) {
                currSum+=n;
            }else {
                currSum=n;
                count++;
                if(count>k) return false;
            }
        }
        return true;
    }
}